packages/flutter/lib/src/rendering/shifted_box.dart

```diff
@@ -91,78 +91,107 @@ class RenderPadding extends RenderShiftedBox {
   ///
   /// The [padding] argument must not be null and must have non-negative insets.
   RenderPadding({
-    @required EdgeInsets padding,
-    RenderBox child
+    @required EdgeInsetsGeometry padding,
+    TextDirection textDirection,
+    RenderBox child,
   }) : assert(padding != null),
        assert(padding.isNonNegative),
+       _textDirection = textDirection,
        _padding = padding,
-       super(child);
+       super(child) {
+    _applyUpdate();
+  }
+
+  // The resolved absolute insets.
+  EdgeInsets _resolvedPadding;
+
+  void _applyUpdate() {
+    final EdgeInsets resolvedPadding = padding.resolve(textDirection);
+    assert(resolvedPadding.isNonNegative);
+    if (resolvedPadding != _resolvedPadding) {
+      _resolvedPadding = resolvedPadding;
+      markNeedsLayout();
+    }
+  }
 
   /// The amount to pad the child in each dimension.
-  EdgeInsets get padding => _padding;
-  EdgeInsets _padding;
-  set padding(EdgeInsets value) {
+  ///
+  /// If this is set to an [EdgeInsetsDirectional] object, then [textDirection]
+  /// must be non-null.
+  EdgeInsetsGeometry get padding => _padding;
+  EdgeInsetsGeometry _padding;
+  set padding(EdgeInsetsGeometry value) {
     assert(value != null);
     assert(value.isNonNegative);
     if (_padding == value)
       return;
     _padding = value;
-    markNeedsLayout();
+    _applyUpdate();
```

Comment:
This will fire twice when changing both padding and text direction.

We could add a function that lets you set the padding and text direction at the same time and use that from the widget to make this more efficient.